core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java

```diff
@@ -26,42 +26,33 @@
 import org.elasticsearch.indices.query.IndicesQueriesRegistry;
 
 import java.io.IOException;
+import java.util.Objects;
 
 public class QueryParseContext {
 
     private static final ParseField CACHE = new ParseField("_cache").withAllDeprecated("Elasticsearch makes its own caching decisions");
     private static final ParseField CACHE_KEY = new ParseField("_cache_key").withAllDeprecated("Filters are always used as cache keys");
 
-    private XContentParser parser;
-    private ParseFieldMatcher parseFieldMatcher = ParseFieldMatcher.EMPTY;
+    private final XContentParser parser;
 
     private IndicesQueriesRegistry indicesQueriesRegistry;
 
-    public QueryParseContext(IndicesQueriesRegistry registry) {
-        this.indicesQueriesRegistry = registry;
+    public QueryParseContext(IndicesQueriesRegistry registry, XContentParser parser, ParseFieldMatcher parseFieldMatcher) {
+        this.indicesQueriesRegistry = Objects.requireNonNull(registry, "indices queries registry cannot be null");
+        this.parser = Objects.requireNonNull(parser, "parser cannot be null");
+        this.parser.setParseFieldMatcher(Objects.requireNonNull(parseFieldMatcher, "parse field matcher cannot be null"));
```

Comment:
ok, I haven't followed all recent changes, I just realized that there are ParseFieldMatcher instances in different places. I also wonder now if it should be in the parser itself, like should it be in the parser or in the parse context? We should pick one, it cannot be both ideally :)